File: security-platform/agents/ios/Sources/SecurityPlatform/Agent.swift
import Foundation
import OpenTelemetryApi
import OpenTelemetrySdk
import ResourceExtension

/// Security Platform Agent for iOS applications.
public class Agent {
    private let config: AgentConfig
    private let redactor: Redactor
    private let policyEngine: PolicyEngine
    private var tracerProvider: TracerProvider?
    private var isStarted: Bool = false

    public init(config: AgentConfig) {
        self.config = config
        self.redactor = Redactor(rules: config.redactionRules)
        self.policyEngine = PolicyEngine(config: config.policy)
    }

    /// Start the agent and initialize OpenTelemetry.
    public func start() throws {
        guard !isStarted else {
            print("Agent already started")
            return
        }

        let otlpEndpoint = config.otlpEndpoint ?? "\(config.controlPlaneURL)/v1/traces"

        // Create resource
        let resource = Resource.getDefault()
            .merging(other: Resource(attributes: [
                "service.name": AttributeValue.string(config.serviceName),
                "deployment.environment": AttributeValue.string(config.environment)
            ]))

        // Create OTLP exporter (simplified - in production use proper OTLP exporter)
        // For now, we'll use stdout exporter as placeholder
        let exporter = StdoutSpanExporter()
        
        let spanProcessor = BatchSpanProcessor(spanExporter: exporter)
        
        self.tracerProvider = TracerProviderBuilder()
            .with(resource: resource)
            .with(spanProcessor: spanProcessor)
            .build()

        OpenTelemetry.registerTracerProvider(tracerProvider: tracerProvider!)
        isStarted = true
        
        print("Security Platform Agent started for service: \(config.serviceName)")
    }

    /// Stop the agent.
    public func stop() {
        guard isStarted else { return }
        tracerProvider?.shutdown()
        isStarted = false
        print("Security Platform Agent stopped")
    }

    /// Redact PII from data.
    public func redact(_ data: Any) -> Any {
        return redactor.redact(data)
    }

    /// Check if an action is allowed by policy.
    public func checkPolicy(action: String, context: [String: Any]) -> PolicyResult {
        return policyEngine.check(action: action, context: context)
    }

    /// Get a tracer for instrumentation.
    public func getTracer(name: String) -> Tracer {
        return OpenTelemetry.instance.tracerProvider.get(instrumentationName: name)
    }
}

File: security-platform/agents/android/src/main/java/com/securityplatform/agent/Agent.kt
package com.securityplatform.agent

import android.content.Context
import io.opentelemetry.api.OpenTelemetry
import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.trace.Tracer
import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter
import io.opentelemetry.sdk.OpenTelemetrySdk
import io.opentelemetry.sdk.resources.Resource
import io.opentelemetry.sdk.trace.SdkTracerProvider
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor
import io.opentelemetry.semconv.ResourceAttributes

/**
 * Security Platform Agent for Android applications.
 * Provides automatic instrumentation and telemetry collection.
 */
class Agent private constructor(private val config: AgentConfig) {
    private var openTelemetry: OpenTelemetry? = null
    private val redactor: Redactor = Redactor(config.redactionRules ?: emptyList())
    private val policyEngine: PolicyEngine = PolicyEngine(config.policy ?: PolicyConfig())
    private var started: Boolean = false

    companion object {
        /**
         * Create a new agent instance with the given configuration.
         */
        @JvmStatic
        fun create(config: AgentConfig): Agent {
            return Agent(config)
        }
    }

    /**
     * Start the agent and initialize OpenTelemetry.
     */
    fun start() {
        if (started) {
            android.util.Log.w(TAG, "Agent already started")
            return
        }

        try {
            val resource = Resource.getDefault()
                .merge(
                    Resource.create(
                        Attributes.of(
                            ResourceAttributes.SERVICE_NAME, config.serviceName ?: "android-app",
                            ResourceAttributes.DEPLOYMENT_ENVIRONMENT, config.environment ?: "production"
                        )
                    )
                )

            val otlpEndpoint = config.otlpEndpoint
                ?: "${config.controlPlaneUrl ?: "https://api.securityplatform.com"}/v1/traces"

            val headers = mutableMapOf<String, String>()
            config.authKey?.let {
                headers["Authorization"] = "Bearer $it"
            }

            val spanExporter = OtlpHttpSpanExporter.builder()
                .setEndpoint(otlpEndpoint)
                .setHeaders(headers)
                .build()

            val tracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(
                    BatchSpanProcessor.builder(spanExporter)
                        .setMaxQueueSize(config.telemetry?.maxQueueSize ?: 2048)
                        .setMaxExportBatchSize(config.telemetry?.batchSize ?: 100)
                        .setExportTimeoutMillis((config.telemetry?.exportTimeoutSeconds ?: 30) * 1000L)
                        .setScheduleDelayMillis((config.telemetry?.batchTimeoutSeconds ?: 5) * 1000L)
                        .build()
                )
                .setResource(resource)
                .build()

            this.openTelemetry = OpenTelemetrySdk.builder()
                .setTracerProvider(tracerProvider)
                .build()

            started = true
            android.util.Log.i(TAG, "Security Platform Agent started successfully")
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to start agent", e)
            throw RuntimeException("Failed to start Security Platform Agent", e)
        }
    }

    /**
     * Stop the agent and shutdown OpenTelemetry.
     */
    fun stop() {
        if (!started) {
            return
        }

        try {
            (openTelemetry as? OpenTelemetrySdk)?.sdkTracerProvider?.shutdown()
            started = false
            android.util.Log.i(TAG, "Security Platform Agent stopped")
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Error stopping agent", e)
        }
    }

    /**
     * Get a tracer for manual instrumentation.
     */
    fun getTracer(instrumentationName: String): Tracer {
        return openTelemetry?.getTracer(instrumentationName)
            ?: throw IllegalStateException("Agent not started")
    }

    /**
     * Redact PII from data.
     */
    fun redact(data: Any?): Any? {
        return redactor.redact(data)
    }

    /**
     * Check if an action is allowed by policy.
     */
    fun checkPolicy(action: String, context: Map<String, Any>? = null): PolicyResult {
        return policyEngine.check(action, context ?: emptyMap())
    }

    private companion object {
        const val TAG = "SecurityPlatformAgent"
    }
}
